### ISSUE TITLE
[Performance] Dev is slow with thousands of route annotations

### ISSUE THREAD

**Issue by pbowyer** [symfony/symfony#49036]
### Symfony version(s) affected

6.2.3

### Description

I am working on a large Symfony app with approx 1700 `@Route` annotations and (so far) 1 `#[Route]` attribute, across 1350 files.

XDebug is disabled.

`bin/console cache:clear` takes 24 seconds to run. 13 seconds of this is spent in `RouterCacheWarmer`.

Almost all this time is spent in `\Symfony\Component\Routing\Loader\AnnotationFileLoader::load()`.

I do not have blackfire so I am limited in the profiling I can do (I have been adding `hrtime()` around calls to see how long they take).

### Why does this matter?
In production it doesn't. In development it's a slowdown in multiple places.

Take a functional test as an example. If I repeatedly run the functional test I pay the cost of the annotations being loaded on the first run. With the fixtures we have, each subsequent run takes ~8 seconds.

If I make any edit to the controller I'm testing, e.g. adding `$x = 1;` inside the method body, the routes are reloaded and it takes ~13 seconds, so 5 seconds longer, to rerun the test.

There is something more complex going on because this adds 5 seconds whereas on `cache:clear` it's a consistent 13 seconds. I have not got to the bottom of that.

### How to reproduce

Add a lot of routes to a project.

### Possible Solution

This project uses DDD/CQRS and subclasses \Symfony\Component\Config\Loader\Loader and for local development I have added a permanent annotation cache outside the Symfony cache directory. This removes the performance drag during development but is not a good solution.

```php
<?php  
declare(strict_types=1);  
  
namespace Example;  
  
use Symfony\Component\Config\Loader\Loader;  
use Symfony\Component\Finder\Finder;  
use Symfony\Component\Finder\SplFileInfo;  
use Symfony\Component\HttpKernel\KernelInterface;  
use Symfony\Component\Routing\RouteCollection;  
  
class DddLoader extends Loader  
{  
    private string $srcDir;  
    public function __construct(KernelInterface $kernel)  
    {  
        $this->srcDir = $kernel->getProjectDir()."/src";  
        // My extra hack
        $this->cacheDir = $kernel->getCacheDir().'/../annotations';  
        @mkdir($this->cacheDir, 0777, true);  
    }  
  
    public function load($resource, string $type = null): RouteCollection  
    {  
        $finder = new Finder();  
        $finder->in($this->srcDir)->name('*Controller.php');  
        $routes = new RouteCollection();
        foreach($finder AS $fileInfo){  
            $hash = md5($fileInfo->getRealPath());  
            if (file_exists($this->cacheDir . "/$hash.php") && filemtime($this->cacheDir . "/$hash.php") > $fileInfo->getMTime()) {  
                $routes->addCollection(unserialize(file_get_contents($this->cacheDir . "/$hash.php")));  
            } else {  
                $annotationLoader = $this->resolver->resolve($fileInfo->getRealPath(), "annotation");    
                if($controllerRoutes = $annotationLoader->load($fileInfo->getRealPath())){  
                    file_put_contents($this->cacheDir . "/$hash.php", serialize($controllerRoutes));  
                    $routes->addCollection($controllerRoutes);  
                }  
            }
        }
        return $routes;  
    }  
  
    public function supports($resource, string $type = null): bool  
    {  
        return ($type == "ddd");  
    }  
}
```

Short of switching from annotations to another format, which I do not have the power to do, are there any improvements that can be made to Symfony to improve performance?

**Comment by nicolas-grekas** [comment]
Can you please provide a reproducer as a git repository that'd be ready for others to play with?
Can you try updating all routes to `#[Route]` (using rector possibly?) and report back if the issue persists?

**Comment by michaljusiega** [comment]
My issue 4 years ago: https://github.com/symfony/symfony/issues/32545

Maybe still related with your problem?

**Comment by derrabus** [comment]
You can also try to back the annotation reader with a filesystem cache of some sort, but I'd really not recommend caching aggressively in dev env. Please try to migrate to attributes. Rector should be able to do the heavy lifting here.

Attributes have the advantage that we don't have to parse doc block strings at runtime anymore. Instead we read information vie reflection that is cached by PHP's own opcache.

But all of this is just wild guesswork. We need a reproducer that we can analyze. Alternatively a Blackfire profile.

**Comment by pbowyer** [comment]
I've got code generation of the Controllers set up at https://github.com/pbowyer/symfony-bug-49036 but failing at the last hurdle to get the demo working with Symfony 6.2.3. I'm getting a `LoaderLoadException`
> [Semantical Error] The class "Symfony\Contracts\Service\Attribute\Required" is not annotated with @Annotation.
Are you sure this class can be used as annotation?
If so, then you need to add @Annotation to the _class_ doc comment of "Symfony\Contracts\Service\Attribute\Required".
If it is indeed no annotation, then you need to add @IgnoreAnnotation("required") to the _class_ doc comment of method Symfony\Bundle\FrameworkBundle\Controller\AbstractController::setContainer() in /path/to/symfony-bug-49036/config/../src/Controller/ (which is being imported from "/path/to/symfony-bug-49036/config/routes.yaml"). Make sure there is a loader supporting the "annotation" type.

I've installed `doctrine/annotations`, changed [`config/routes.yaml`](https://github.com/pbowyer/symfony-bug-49036/blob/main/config/routes.yaml) to the old form - what have I missed?

**Comment by stof** [comment]
@pbowyer your error is the issue https://github.com/symfony/symfony/issues/48792 which is already fixed but not released yet. As a workaround, force installing `doctrine/annotations` 1.x instead of 2.0

**Comment by pbowyer** [comment]
@stof thanks, that fixed it and the reproducer repository is working.

## Time to run `cache:clear` with 1700 routes
With old-school annotations: 12.4s
With PSR-4 routes: 13.5s
With attributes: 12.3s

For PSR-4 routes I replaced `config/routes.yaml` with
```
controllers:
    resource:
        path: ../src/Controller/
        namespace: App\Controller
    type: annotation
```
Tests run on a M1 MBP using `PHP 8.1.14 (cli) (built: Jan 12 2023 02:41:00) (NTS)`

*With 2500 routes, the time for old-school annotations jumps to ~21s. However `RouterCacheWarmer` is only 5s of this. Hopefully someone who has Blackfire access can see why*

## Time to reload a page after modifying the controller action with 1700 routes

> If I make any edit to the controller I'm testing, e.g. adding $x = 1; inside the method body, the routes are reloaded and it takes ~13 seconds, so 5 seconds longer, to rerun the test.

I haven't been able to reproduce this magnitude of slowdown in my demo. Locally I'm using `symfony serve` to deliver the test page.

I request a page such as `time wget http://127.0.0.1:8000/foo3/bar13/baz53/33 -O /dev/null`.

After adding a single variable to a controller action, the next refresh (normal refresh) was:

With old-school annotations: 650ms (151ms)
With PSR-4 routes: 1150ms (290ms) -- no idea what I did wrong
With attributes: 640ms (153ms)

**Comment by nicolas-grekas** [comment]
Thanks for the reproducer. This is unrelated to the cache warmer. The issue is already fixed by #48802.

Upgrading to v6.3 gives a 44% improvement. Thanks Blackfire.io for cutting the time to understand such issues ;)

![image](https://user-images.githubusercontent.com/243674/215069491-0a2a98c7-a4c6-4c93-84d6-98a01e573414.png)

**Comment by pbowyer** [comment]
Thanks @nicolas-grekas I can't wait!
